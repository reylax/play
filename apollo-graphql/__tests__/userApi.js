require('dotenv').config()
const { startTestServer, setTestClient, toPromise } = require('./__utils')
const jwt = require('jsonwebtoken')
const jwtPrivateKey = process.env.jwtPrivateKey
const { redis } = require('../src/redis')
const {
  changePassword,
  updateUser,
  refreshToken,
  logIn,
  signUp,
  getUserInfo,
  requestEmail,
} = require('./__query')

var executeOperation, conn, httpServer

beforeAll( async () => {
  const testServer = await startTestServer({
    port: 4001,
    databaseName: "common",
  });
  conn = testServer.conn
  httpServer = testServer.httpServer
  executeOperation = setTestClient({port: 4001, token: ""})
});

afterAll( async () => {
  await conn.dropDatabase()
  httpServer.close()
});

describe('sign up', () => {
  test('sign up and wait database create index', async () => {
    const res = await toPromise(executeOperation({
      query:signUp, 
      variables: {
        username: "hello",
        email: "sayhimage@outlook.com",
        password: "correctPassword",
      }
    }))
    await new Promise(resolve => setTimeout(resolve, 2000))
    expect(res.data).toMatchSnapshot();
    expect(res.data.signUp.status).toBe(201)
  });
  test('with same name', async () => {
    const res = await toPromise(executeOperation({
      query:signUp, 
      variables: {
        username: "hello",
        email: "sayhimage@outlookk.com",
        password: "correctPassword",
      }
    }))
    expect(res.data).toMatchSnapshot();
    expect(res.data.signUp.status).toBe(400)
  })
  test('with same email', async () => {
    const res = await toPromise(executeOperation({
      query:signUp, 
      variables: {
        username: "halo",
        email: "sayhimage@outlook.com",
        password: "correctPassword",
      }
    }))
    expect(res.data).toMatchSnapshot();
    expect(res.data.signUp.status).toBe(400)
  })
  test('username too long', async () => {
    var res = await toPromise(executeOperation({
      query:signUp, 
      variables: {
        username: "q01234567890123456789qqq",
        email: "sayhimage@outlookk.com",
        password: "fjkdfjfklds",
      }}))
      expect(res.data).toMatchSnapshot()
      expect(res.data.signUp.status).toBe(400)
  })
  test('username not start with letter', async () => {
    var res = await toPromise(executeOperation({
      query:signUp, 
      variables: {
        username: "01234567890123456789",
        email: "sayhimage@outlookk.com",
        password: "fjkdfjfklds",
      }}))
      expect(res.data).toMatchSnapshot()
      expect(res.data.signUp.status).toBe(400)
  })
  test('username too short ', async () => {
    var res = await toPromise(executeOperation({
      query:signUp, 
      variables: {
        username: "U012",
        email: "sayhimage@outlookk.com",
        password: "fjkdfjfklds",
      }}))
      expect(res.data).toMatchSnapshot()
      expect(res.data.signUp.status).toBe(400)
  })
  test('invalid email', async () => {
    var res = await toPromise(executeOperation({
      query:signUp, 
      variables: {
        username: "U012",
        email: "sayhimage@outlook.",
        password: "fjkdfjfklds",
      }}))
      expect(res.data).toMatchSnapshot()
      expect(res.data.signUp.status).toBe(400)
  })
})

describe('log in with username', () => {
  test('correct Info', async () => {
    var res = await toPromise(executeOperation({
      query:logIn, 
      variables: {
        key: "username",
        value: "hello",
        password: "correctPassword",
      }}))
    var { status, message } = res.data.login
    expect(message).toHaveLength(240)
    const user = jwt.verify(message, jwtPrivateKey)
    expect(user.username).toBe("hello")
    expect(status).toBe(201)
    expect(res.data.login).toMatchSnapshot()
  })
  test('wrong password', async () => {
    res = await toPromise(executeOperation({
      query:logIn, 
      variables: {
        key: "username",
        value: "hello",
        password: "wrongPassword",
      }}))
    expect(res.data).toMatchSnapshot()
    expect(res.data.login.status).toBe(400)
  })
  test('bad username', async () => {
    res = await toPromise(executeOperation({
      query:logIn, 
      variables: {
        key: "username",
        value: "helloo",
        password: "wrongPassword",
      }}))
    expect(res.data).toMatchSnapshot()
    expect(res.data.login.status).toBe(400)
  }) 
})

describe('log in with email', () => {
  test('correct Info', async () => {
    var res = await toPromise(executeOperation({
      query:logIn, 
      variables: {
        key: "email",
        value: "sayhimage@outlook.com",
        password: "correctPassword",
      }}))
    var status = res.data.login.status
    var message = res.data.login.message
    expect(message).toHaveLength(240)
    const user = jwt.verify(message, jwtPrivateKey)
    expect(user.username).toBe("hello")
    expect(status).toBe(201)
    expect(res.data.login).toMatchSnapshot()
  })
  test('wrong password', async () => {
    res = await toPromise(executeOperation({
      query:logIn, 
      variables: {
        key: "email",
        value: "sayhimage@outlook.com",
        password: "wrongPassword",
      }}))
    expect(res.data).toMatchSnapshot()
    expect(res.data.login.status).toBe(400)
  })
  test('wrong email', async () => {
    res = await toPromise(executeOperation({
      query:logIn, 
      variables: {
        key: "email",
        value: "sayhimage@outloook.com",
        password: "correctPassword",
      }}))
    expect(res.data).toMatchSnapshot()
    expect(res.data.login.status).toBe(400)
  })
})

describe('refresh token', () => {
  var token = ""
  var newToken = ""
  test('login get token', async () => {
    var res = await toPromise(executeOperation({
      query:logIn, 
      variables: {
        key: "username",
        value: "hello",
        password: "correctPassword",
      }}))
    var status = res.data.login.status
    token = res.data.login.message
    expect(token).toHaveLength(240)
    var user = jwt.verify(token, jwtPrivateKey)
    expect(user.username).toBe("hello")
    expect(status).toBe(201)
    expect(res.data).toMatchSnapshot()
  })

  test('refresh new token', async () => {
    executeOperation = setTestClient({token: token, port: 4001})
    await new Promise(resolve => setTimeout(resolve, 1000))
    const res = await toPromise(executeOperation({
      query: refreshToken,
    }))
    newToken = res.data.refreshToken.message
    expect(res.data.refreshToken.status).toBe(201)
    expect(newToken).toHaveLength(240)
    expect(newToken === token).toBe(false)
    user = jwt.verify(newToken, jwtPrivateKey)
    expect(user.username).toBe("hello")
    expect(res.data).toMatchSnapshot()
  })

  test('get self only info with new token', async () => {
    executeOperation = setTestClient({token: newToken, port: 4001})
    res = await toPromise(executeOperation({
      query: getUserInfo,
      variables: {
        username: 'hello',
      }
    }))
    expect(res.data.userInfo.email).toBe("sayhimage@outlook.com")
    expect(res.data).toMatchSnapshot()
  })
})

describe('update userInfo', () => {
  var newInfo = {
    username: "newUsername",
    avatar: "avatar",
    backgroundImage: "backgroundImage",
    website: "nah yet",
    bio: "hello world",
    theme: {
      primary: "blueHeart",
      background: "dim",
      language: "english",
    }
  }
  newInfo = JSON.stringify(newInfo)
  test('without login', async () => {
      executeOperation = setTestClient({token: "", port: 4001})
      const res = await toPromise(executeOperation({
        query:updateUser, 
        variables: {
          newInfo,
        }
      }))
      expect(res.data).toMatchSnapshot();
      expect(res.data.updateUser.status).toBe(400)
  })
  test('with login expect seccess', async () => {
    var res = await toPromise(executeOperation({
      query:logIn, 
      variables: {
        key: "username",
        value: "hello",
        password: "correctPassword",
      }}))
    const token = res.data.login.message
    expect(res.data).toMatchSnapshot();
    executeOperation = setTestClient({token: token, port: 4001})
    res = await toPromise(executeOperation({
      query:updateUser, 
      variables: {
        newInfo,
      }
    }))
    expect(res.data).toMatchSnapshot();
    expect(res.data.updateUser.status).toBe(202)
    
  })
})

describe('request email', () => {
  test('with correct argument expect success', async () => {
    const res = await toPromise(executeOperation({
      query: requestEmail, 
      variables: {
        purpose: "passwordReset",
        email: "sayhimage@outlook.com",
        key: process.env.sentEmailKey,
      }
    }))
    expect(res.data).toMatchSnapshot()
    expect(res.data.requestEmail.status).toBe(201)
    const code = await redis.hget('passwordReset', "sayhimage@outlook.com")
    expect(code).toHaveLength(6)
  })

  test('with invalid purpose', async () => {
    const res = await toPromise(executeOperation({
      query: requestEmail, 
      variables: {
        purpose: "just sent",
        email: "sayhimage@outlook.com",
        key: process.env.sentEmailKey,
      }
    }))
    expect(res.data).toMatchSnapshot()
    expect(res.data.requestEmail.status).toBe(400)

  })

  test('with invalid key', async () => {
    const res = await toPromise(executeOperation({
      query: requestEmail, 
      variables: {
        purpose: "passwordReset",
        email: "sayhimage@outlook.com",
        key: "fjkslfjkl",
      }
    }))
    expect(res.data).toMatchSnapshot()
    expect(res.data.requestEmail.status).toBe(400)
  })
  test('with invalid email', async () => {
    const res = await toPromise(executeOperation({
      query: requestEmail, 
      variables: {
        purpose: "passwordReset",
        email: "sayhimage@.com",
        key: process.env.sentEmailKey,
      }
    }))
    expect(res.data).toMatchSnapshot()
    expect(res.data.requestEmail.status).toBe(400)
  })
})

describe('change password', () => {
  test('sign up', async () => {
    const res = await toPromise(executeOperation({
      query:signUp, 
      variables: {
        username: "halo",
        email: "sayhaimage@outlook.com",
        password: "correctPassword",
      }
    }))
    expect(res.data).toMatchSnapshot();
    expect(res.data.signUp.status).toBe(201)
  })
  test('get code', async () => {
    const res = await toPromise(executeOperation({
      query: requestEmail, 
      variables: {
        email: "sayhaimage@outlook.com",
        key: process.env.sentEmailKey,
        purpose: "passwordReset",
    }}))
    expect(res.data).toMatchSnapshot()
    expect(res.data.requestEmail.status).toBe(201)
    const code = await redis.hget('passwordReset', "sayhaimage@outlook.com")
    expect(code).toHaveLength(6)
  })
  test('do change', async () => {
    const code = await redis.hget('passwordReset', "sayhaimage@outlook.com")
    res = await toPromise(executeOperation({
      query:changePassword, 
      variables: {
        email: "sayhaimage@outlook.com",
        cert: code,
        newPassword: "ILoveSpaceX"
      }}))
    expect(res.data.changePassword.status).toBe(202)
    expect(res.data).toMatchSnapshot()
  })
  test('log in with new password', async () => {
    res = await toPromise(executeOperation({
      query:logIn, 
      variables: {
        key: "username",
        value: "halo",
        password: "ILoveSpaceX",
      }}))
    expect(res.data).toMatchSnapshot()
    expect(res.data.login.status).toBe(201)
  })
  test('login with bad password', async () => {
    res = await toPromise(executeOperation({
      query:logIn, 
      variables: {
        key: "username",
        value: "halo",
        password: "badpassword",
      }}))
    expect(res.data).toMatchSnapshot()
    expect(res.data.login.status).toBe(400)
  });
})









