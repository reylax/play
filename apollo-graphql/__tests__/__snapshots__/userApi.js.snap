// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`change password do change 1`] = `
Object {
  "changePassword": Object {
    "message": "ok",
    "status": 202,
  },
}
`;

exports[`change password get code 1`] = `
Object {
  "requestEmail": Object {
    "message": "code have sent",
    "status": 201,
  },
}
`;

exports[`change password log in with new password 1`] = `
Object {
  "login": Object {
    "message": "eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI2MTAyNzU4YjU5ZjMzMTM3YjAzODUwYzUiLCJ1c2VybmFtZSI6ImhhbG8iLCJpYXQiOjE2Mjc1NTExMTYsImV4cCI6MTYzMDE0MzExNn0.qHfpyRhNMeVpAtuoSM7sQMoxGSq8o8rKFkdpwnXc7OXiZr609YTLN0mbQQQqF2QuxcjpDT0xg37iXci7kAeyDA",
    "status": 201,
  },
}
`;

exports[`change password login with bad password 1`] = `
Object {
  "login": Object {
    "message": "password or username not correct",
    "status": 400,
  },
}
`;

exports[`change password sign up 1`] = `
Object {
  "signUp": Object {
    "message": "success",
    "status": 201,
  },
}
`;

exports[`log in with email correct Info 1`] = `
Object {
  "message": "eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI2MTAyNzU4ODU5ZjMzMTM3YjAzODUwYjMiLCJ1c2VybmFtZSI6ImhlbGxvIiwiaWF0IjoxNjI3NTUxMTE0LCJleHAiOjE2MzAxNDMxMTR9.UD2rjr0PKgMu_gLuhtZs4pdfjCACM2w7NlcY17JVqSp38BJN56UWhUN5YPKDIqupaEBBV5qqb7yRNcaiYHvgNw",
  "status": 201,
}
`;

exports[`log in with email wrong email 1`] = `
Object {
  "login": Object {
    "message": "password or username not correct",
    "status": 400,
  },
}
`;

exports[`log in with email wrong password 1`] = `
Object {
  "login": Object {
    "message": "password or username not correct",
    "status": 400,
  },
}
`;

exports[`log in with username bad username 1`] = `
Object {
  "login": Object {
    "message": "password or username not correct",
    "status": 400,
  },
}
`;

exports[`log in with username correct Info 1`] = `
Object {
  "message": "eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI2MTAyNzU4ODU5ZjMzMTM3YjAzODUwYjMiLCJ1c2VybmFtZSI6ImhlbGxvIiwiaWF0IjoxNjI3NTUxMTE0LCJleHAiOjE2MzAxNDMxMTR9.UD2rjr0PKgMu_gLuhtZs4pdfjCACM2w7NlcY17JVqSp38BJN56UWhUN5YPKDIqupaEBBV5qqb7yRNcaiYHvgNw",
  "status": 201,
}
`;

exports[`log in with username wrong password 1`] = `
Object {
  "login": Object {
    "message": "password or username not correct",
    "status": 400,
  },
}
`;

exports[`refresh token get self only info with new token 1`] = `
Object {
  "userInfo": Object {
    "email": "sayhimage@outlook.com",
    "username": "hello",
  },
}
`;

exports[`refresh token login get token 1`] = `
Object {
  "login": Object {
    "message": "eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI2MTAyNzU4ODU5ZjMzMTM3YjAzODUwYjMiLCJ1c2VybmFtZSI6ImhlbGxvIiwiaWF0IjoxNjI3NTUxMTE0LCJleHAiOjE2MzAxNDMxMTR9.UD2rjr0PKgMu_gLuhtZs4pdfjCACM2w7NlcY17JVqSp38BJN56UWhUN5YPKDIqupaEBBV5qqb7yRNcaiYHvgNw",
    "status": 201,
  },
}
`;

exports[`refresh token refresh new token 1`] = `
Object {
  "refreshToken": Object {
    "message": "eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI2MTAyNzU4ODU5ZjMzMTM3YjAzODUwYjMiLCJ1c2VybmFtZSI6ImhlbGxvIiwiaWF0IjoxNjI3NTUxMTE1LCJleHAiOjE2MzAxNDMxMTV9.mz6NxByTkAs4dmZQKJxeTwt-MTH-k5fw-_GaccGhr96B0uWqll5dqOFLi7TWXakUqvX0tOH_e6qjM7UspiYV7g",
    "status": 201,
  },
}
`;

exports[`request email with correct argument expect success 1`] = `
Object {
  "requestEmail": Object {
    "message": "code have sent",
    "status": 201,
  },
}
`;

exports[`request email with invalid email 1`] = `
Object {
  "requestEmail": Object {
    "message": "invalid email",
    "status": 400,
  },
}
`;

exports[`request email with invalid key 1`] = `
Object {
  "requestEmail": Object {
    "message": "invalid key",
    "status": 400,
  },
}
`;

exports[`request email with invalid purpose 1`] = `
Object {
  "requestEmail": Object {
    "message": "invalid purpose",
    "status": 400,
  },
}
`;

exports[`sign up invalid email 1`] = `
Object {
  "signUp": Object {
    "message": "invalid information",
    "status": 400,
  },
}
`;

exports[`sign up sign up and wait database create index 1`] = `
Object {
  "signUp": Object {
    "message": "success",
    "status": 201,
  },
}
`;

exports[`sign up username not start with letter 1`] = `
Object {
  "signUp": Object {
    "message": "invalid information",
    "status": 400,
  },
}
`;

exports[`sign up username too long 1`] = `
Object {
  "signUp": Object {
    "message": "invalid information",
    "status": 400,
  },
}
`;

exports[`sign up username too short  1`] = `
Object {
  "signUp": Object {
    "message": "invalid information",
    "status": 400,
  },
}
`;

exports[`sign up with same email 1`] = `
Object {
  "signUp": Object {
    "message": "email must be unique",
    "status": 400,
  },
}
`;

exports[`sign up with same name 1`] = `
Object {
  "signUp": Object {
    "message": "username must be unique",
    "status": 400,
  },
}
`;

exports[`update userInfo with login expect seccess 1`] = `
Object {
  "login": Object {
    "message": "eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI2MTAyNzU4ODU5ZjMzMTM3YjAzODUwYjMiLCJ1c2VybmFtZSI6ImhlbGxvIiwiaWF0IjoxNjI3NTUxMTE1LCJleHAiOjE2MzAxNDMxMTV9.mz6NxByTkAs4dmZQKJxeTwt-MTH-k5fw-_GaccGhr96B0uWqll5dqOFLi7TWXakUqvX0tOH_e6qjM7UspiYV7g",
    "status": 201,
  },
}
`;

exports[`update userInfo with login expect seccess 2`] = `
Object {
  "updateUser": Object {
    "message": "ok",
    "status": 202,
  },
}
`;

exports[`update userInfo without login 1`] = `
Object {
  "updateUser": Object {
    "message": "invalid quest",
    "status": 400,
  },
}
`;
